lib/more_collections.nit

```diff
@@ -193,3 +193,315 @@ class DefaultMap[K, V]
+		end
+		length += 1
+	end
+
+	redef fun unshift(item)
+	do
+		var node = head_node
+		if not node.full then
+			if node.head_index > 0 then
+				# There's room at the head
+				node.head_index -= 1
+			else
+				# Move everything over by `d`
+				assert node.tail_index < node.capacity
+				var d = (node.capacity-node.tail_index) / 2 + 1
+				for i in [node.tail_index-1..node.head_index].step(-1) do
+					node.items[i+d] = node.items[i]
+				end
+				for i in d.times do node.items[node.head_index+i] = null
+				node.head_index += d-1
+				node.tail_index += d
+			end
+			node.items[node.head_index] = item
+		else
+			# New node!
+			node = new UnrolledNode[E](nodes_length)
+			head_node.prev = node
+			node.next = head_node
+			head_node = node
+			node.head_index = node.capacity-1
+			node.tail_index = node.capacity
+			node.items[node.capacity-1] = item
+		end
+		length += 1
+	end
+
+	redef fun pop
+	do
+		assert not_empty
+
+		var node = tail_node
+		while node.length == 0 do
+			# Delete empty node
+			var nullable_node = node.prev
+			assert is_not_empty: nullable_node != null
+			node = nullable_node
+			node.next = null
+			self.tail_node = node
+		end
+
+		var item = node.items[node.tail_index-1]
+		node.tail_index -= 1
+		length -= 1
+		return item
+	end
+
+	redef fun shift
+	do
+		assert not_empty
+
+		var node = head_node
+		while node.length == 0 do
+			# Delete empty node
+			var nullable_node = node.next
+			assert is_not_empty: nullable_node != null
+			node = nullable_node
+			node.prev = null
+			self.head_node = node
+		end
+
+		var item = node.items[node.head_index]
+		node.head_index += 1
+		length -= 1
+		return item
+	end
+
+	redef fun insert(item, index)
+	do
+		if index == length then
+			push item
+			return
+		end
+
+		var node = node_at(index)
+		index = index_within_node
+		if node.full then
+			# Move half to a new node
+			var new_node = new UnrolledNode[E](nodes_length.max(node.capacity))
+
+			# Plug in the new node
+			var next_node = node.next
+			if next_node == null then
+				self.tail_node = new_node
+			else next_node.prev = new_node
+
+			new_node.next = next_node
+			node.next = new_node
+			new_node.prev = node
+
+			# Move items at and after `index` to the new node
+			var to_displace = node.length-index
+			var offset = (new_node.capacity-to_displace)/2
+			for i in [0..to_displace[ do
+				new_node.items[offset+i] = node.items[index+i]
+				node.items[index+i] = null
+			end
+			new_node.head_index = offset
+			new_node.tail_index = offset + to_displace
+			node.tail_index -= to_displace
+
+			# Store `item`
+			if index > node.capacity / 2 then
+				new_node.items[offset-1] = item
+				new_node.head_index -= 1
+			else
+				node.items[node.head_index+index] = item
+				node.tail_index += 1
+			end
+		else
+			if node.tail_index < node.capacity then
+				# Move items towards the tail
+				for i in [node.tail_index-1..node.head_index+index].step(-1) do
+					node.items[i+1] = node.items[i]
```

Comment:
I'm a little tired to see these loops for copying You should delegate the responsibility of the copy to the nodes I think, this may cleanup and factorize code